src/integrations/watchers/indexable-permalink-watcher.php

```diff
@@ -0,0 +1,166 @@
+use Yoast\WP\SEO\WordPress\Wrapper;
+
+/**
+ * Handles updates to the permalink_structure for the Indexables table.
+ */
+class Indexable_Permalink_Watcher implements Integration_Interface {
+	/**
+	 * @var Post_Type_Helper
+	 */
+	private $post_type;
+
+	/**
+	 * @inheritdoc
+	 * @codeCoverageIgnore
+	 */
+	public static function get_conditionals() {
+		return [ Migrations_Conditional::class ];
+	}
+
+	/**
+	 * Indexable_Permalink_Watcher constructor.
+	 *
+	 * @codeCoverageIgnore
+	 *
+	 * @param Post_Type_Helper $post_type The post type helper.
+	 */
+	public function __construct( Post_Type_Helper $post_type ) {
+		$this->post_type = $post_type;
+	}
+
+	/**
+	 * @inheritDoc
+	 * @codeCoverageIgnore
+	 */
+	public function register_hooks() {
+		\add_action( 'update_option_permalink_structure', [ $this, 'reset_permalinks' ] );
+		\add_action( 'update_option_category_base', [ $this, 'reset_permalinks_term' ], 10, 3 );
+		\add_action( 'update_option_tag_base', [ $this, 'reset_permalinks_term' ], 10, 3 );
+	}
+
+	/**
+	 * Resets the permalinks for everything that is related to the permalink structure.
+	 */
+	public function reset_permalinks() {
+		$post_types = $this->get_post_types();
+		foreach ( $post_types as $post_type ) {
+			$this->reset_permalinks_post_type( $post_type );
+		}
+
+		$taxonomies = $this->get_taxonomies_for_post_types( $post_types );
+		foreach ( $taxonomies as $taxonomy ) {
+			$this->reset_permalink_indexables( 'term', $taxonomy );
+		}
+
+		$this->reset_permalink_indexables( 'user' );
+		$this->reset_permalink_indexables( 'date-archive' );
+		$this->reset_permalink_indexables( 'system-page' );
+	}
+
+	/**
+	 * Resets the permalink for the given post type.
+	 *
+	 * @param string $post_type The post type to reset.
+	 */
+	public function reset_permalinks_post_type( $post_type ) {
+		$this->reset_permalink_indexables( 'post', $post_type );
+		$this->reset_permalink_indexables( 'post-type-archive', $post_type );
+	}
+
+	/**
+	 * Resets the term indexables when the base has been changed.
+	 *
+	 * @param string $old  Unused. The old option value.
+	 * @param string $new  Unused. The new option value.
+	 * @param string $type The option name.
+	 */
+	public function reset_permalinks_term( $old, $new, $type ) {
+		$subtype = $type;
+
+		// When the subtype contains _base, just strip it.
+		if ( strstr( $subtype, '_base' ) ) {
+			$subtype = substr( $type, 0, -5 );
+		}
+
+		$this->reset_permalink_indexables( 'term', $subtype );
+	}
+
+	/**
+	 * Retrieves a list with the public post types.
+	 *
+	 * @codeCoverageIgnore
+	 *
+	 * @return array The post types.
+	 */
+	protected function get_post_types() {
+		/**
+		 * Filters the post types.
+		 *
+		 * @param array  $post_types The post type names.
+		 * @param string $context    The context where it happens.
+		 *
+		 * @return array The post types.
+		 */
+		$post_types = \apply_filters( 'wpseo_post_types', $this->post_type->get_public_post_types(), 'update_permalink' );
```

Comment:
I am not sure about adding the context where the filter is called as a separate argument. 
Having the context separately adds an extra documentation burden whenever new context variables are added. Adding it to the name, e.g. `wpseo_post_types_update_permalink` would be better in my opinion.